apps/omg_watcher/test/block_getter/core_test.exs

```diff
@@ -642,6 +642,130 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
+
+    {[{%Block{number: 2_000}, 1}, {%Block{number: 3_000}, 3}], 2, _, _} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 2_000, eth_height: 1}, %{blknum: 3_000, eth_height: 3}],
+        3
+      )
+  end
+
+  test "BlockGetter updates mappings of eth_height done by blknum" do
+    state =
+      init_state(synced_height: 1, start_block_number: 1000)
+      |> Core.get_numbers_of_blocks_to_download(5_000)
+      |> assert_check([2_000, 3_000, 4_000])
+      |> handle_downloaded_block(%Block{number: 3_000})
+
+    {[], 1, [], %Core{eth_height_done_by_blknum: %{2_000 => 2, 3_000 => 3}} = state} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 2_000, eth_height: 2}, %{blknum: 3_000, eth_height: 3}],
+        3
+      )
+
+    # if block moves forward it is completely valid
+    {[], 1, [], %Core{eth_height_done_by_blknum: %{2_000 => 2, 3_000 => 3}} = state} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 2_000, eth_height: 3}, %{blknum: 3_000, eth_height: 3}],
+        3
+      )
+
+    # if block moves backward we must update the mapping so we never go backward
+    {[], 1, [], %Core{eth_height_done_by_blknum: %{2_000 => 1, 3_000 => 3}}} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 2_000, eth_height: 1}, %{blknum: 3_000, eth_height: 3}],
+        3
+      )
+  end
+
+  test "already applied cc block appears in a block above synced_height (due to a reorg)" do
+    state =
+      init_state()
+      |> Core.get_numbers_of_blocks_to_download(5_000)
+      |> assert_check([1_000, 2_000, 3_000, 4_000])
+      |> handle_downloaded_block(%Block{number: 1_000})
+
+    {[{%Block{number: 1_000}, 2}], 1, [], %Core{eth_height_done_by_blknum: %{1_000 => 2, 2_000 => 3}} = state} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 1_000, eth_height: 2}, %{blknum: 2_000, eth_height: 3}],
+        3
+      )
+
+    {state, 2, _} = Core.apply_block(state, 1_000)
+    assert %{2000 => 3} == state.eth_height_done_by_blknum
+
+    state = state |> handle_downloaded_block(%Block{number: 2_000})
+
+    {[{%Block{number: 2_000}, 3}], 2, [], state} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 1_000, eth_height: 3}, %{blknum: 2_000, eth_height: 3}],
+        3
+      )
+
+    assert %{2_000 => 3} == state.eth_height_done_by_blknum
+  end
+
+  test "apply block with eth_height lower than synced_height" do
+    state =
+      init_state(synced_height: 2)
+      |> Core.get_numbers_of_blocks_to_download(2_000)
+      |> assert_check([1_000])
+      |> handle_downloaded_block(%Block{number: 1_000})
+
+    {[{%Block{number: 1_000}, 1}], 2, [], state} =
+      Core.get_blocks_to_apply(
+        state,
+        [%{blknum: 1_000, eth_height: 1}],
+        3
+      )
+
+    {_, 2, _} = Core.apply_block(state, 1_000)
+  end
+
+  test "move forward even though an applied block appears in submissions" do
```

Comment:
:+1: on the thoroughness of tests, Just tests for the `get_eth_range_for_block_submitted_events` part need adding.